### PR TITLE
ENH: Add mesh registration via point sets

### PR DESCRIPTION
**Pull request by tbirdso**
Adds Python classes for mesh-to-mesh registration via point set entropy metrics. Includes pytest and examples coverage of new functionality.

Depends on #39 .

### CODE REVIEW

File: src/registration/diffeoregistrar.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
+# Filename:     DiffeoRegistrar.py
+# Contributors: Tom Birdsong
+# Purpose:  Python implementation of Slicer SALT mesh-to-mesh registration Cxx CLI.
+#    Ported from https://github.com/slicersalt/RegistrationBasedCorrespondence/
+
+#!/usr/bin/env python3
```
Comment (thewtex) [inline src/registration/diffeoregistrar.py:6]: This has to be the first line in the file.

File: src/registration/diffeoregistrar.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                                                DisplacementFieldType]
+    DisplacementFieldTransformType = \
+      itk.DisplacementFieldTransform[itk.F, Dimension]
+
+    DiffeoTransformFilterType = \
+      itk.TransformMeshFilter[MeshType,
+                              MeshType,
+                              itk.Transform[itk.F,Dimension,Dimension]]
+
+    def __init__(self):
```
Comment (thewtex) [inline src/registration/diffeoregistrar.py:50]: Please set up the editor to use 4 space indent per Python standard style.

File: src/registration/pointsetentropyregistrar.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    RegistrationType = itk.ImageRegistrationMethodv4[ImageType,ImageType]#,TransformType, ImageType, PointSetType ]
+    TransformFilterType = itk.TransformMeshFilter[MeshType,MeshType,itk.Transform[itk.F,Dimension,Dimension]]
+
+    def __init__(self):
+      pass
+  
+    def register(self, fixedMesh:MeshType, movingMesh:MeshType,
+                  outFile:str=None, verbose=False) -> MeshType:
+      fixedPointSet = self.PointSetType.New(Points=fixedMesh.GetPoints())
+      movingPointSet = self.PointSetType.New(Points=movingMesh.GetPoints())
```
Comment (thewtex) [inline src/registration/pointsetentropyregistrar.py:49]: We should use snake_case for idiomatic Python instance variable naming, `movingPointSet` -> `moving_point_set`.

**Review by thewtex — COMMENTED** [review]
Looking great!

**Review by dzenanz — CHANGES_REQUESTED** [review]
We shouldn't add large data files directly into a repository. These meshes (`*-mesh.vtk`) probably qualify as large. See [ITK's data documentation](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Documentation/Data.md) for alternative approach.

**Comment by dzenanz** [comment]
Also please rebase on current master. I just merged the other PR.

**Comment by tbirdso** [comment]
Rebase complete.

> See [ITK's data documentation](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Documentation/Data.md) for alternative approach.

Would there be any issues with using content links to the publicly accessible [Girder directory](https://data.kitware.com/#collection/5dcc6691e3566bda4b802172/folder/5f9daa3750a41e3d1924d870) of MicroCT meshes?

**Comment by dzenanz** [comment]
> Would there be any issues with using content links to the publicly accessible Girder directory of MicroCT meshes?

Those meshes are kind of large (50MB-ish) for unit test data. You should put those smaller (1.5MB-ish) meshes into the existing [Girder directory for HASI test data](https://data.kitware.com/#collection/57b5c9e58d777f126827f5a1/folder/5ef371d19014a6d84edb5f48).

**Comment by tbirdso** [comment]
Where can I find the smaller meshes? Am I overlooking them in the (microCT directory)[https://data.kitware.com/#collection/5dcc6691e3566bda4b802172/folder/5f9daa3750a41e3d1924d870]?

Also, the meshes will be used in both examples and tests which are contained in different folders. Is it better practice to download meshes twice or have one module reference the other's data, i.e. set the example notebook to reference test data?

**Comment by tbirdso** [comment]
Also -- should I be concerned about check failures? I'm getting test failures when I build locally as well, but it seems like these are present in previous commits. This PR does not make any changes to cxx tests.

**Comment by dzenanz** [comment]
> Where can I find the smaller meshes?

Uh-uh, I now realize that you tried to add meshes with 1.5M lines, not of 1.5MB 😄 In that case, going with the [full-resolution meshes](https://data.kitware.com/#collection/5dcc6691e3566bda4b802172/folder/5f9daa3750a41e3d1924d870) is fine.

> Also -- should I be concerned about check failures?

No. Some tests are running out of memory, and there are some failures due to GTest crashes. It all existed before this PR.

> Is it better practice to download meshes twice?

I am not sure whether we can avoid it. If we can, that would be slightly better. But don't spent much time trying to accomplish that.

**Comment by tbirdso** [comment]
Since Pytests do not need to be built by CMake and we're seeing unrelated issues in building Cxx tests, in the most recent commit I relied on the `urllib.request` module to download meshes at runtime as the example notebook `MouseFemureMicroCT.ipynb` already does.

**Comment by tbirdso** [comment]
Rebased on `master`.

**Comment by tbirdso** [comment]
@thewtex @vicory I've added an initial implementation of VTK k-d tree mesh resampling in the most recent commit using `numpy` arrays to convert between an ITK mesh and VTK polydata object (thanks @vicory for the base algorithm). I noticed that the `SPHARM-PDM` repo we previously referenced for `ITKMeshProcrustesAlignFilter` implements [`itkMeshTovtkPolyDataFilter`](https://github.com/NIRALUser/SPHARM-PDM/blob/master/Modules/CLI/MetaMeshTools/itkMeshToVTKPolyDataFilter.h), [`itkMeshTovtkPolyData`](https://github.com/NIRALUser/SPHARM-PDM/blob/master/Modules/CLI/MetaMeshTools/itkMeshTovtkPolyData.h) and [`vtkPolyDataToitkMesh`](https://github.com/NIRALUser/SPHARM-PDM/blob/master/Modules/CLI/MetaMeshTools/vtkPolyDataToitkMesh.h) classes. These do extra work of recreating cells in the VTK polydata objects that are unused in template resampling, but would avoid the overhead of relying on `numpy` arrays.

Would it be worthwhile to look at wrapping these scripts as part of the `hasi` Python module?

**Comment by dzenanz** [comment]
Tom, maybe take a look at https://github.com/InsightSoftwareConsortium/ITKMeshToPolyData.

**Comment by tbirdso** [comment]
Thanks @dzenanz , I believe [`ITKMeshToPolyData`](https://github.com/InsightSoftwareConsortium/ITKMeshToPolyData) is also used in [`ITKWidgets`](https://github.com/InsightSoftwareConsortium/itkwidgets) so it also has the advantage that we're implicitly including it as a dependency for the `hasi` module. It looks like it's a one-way conversion right now -- is there a similar `ITKMeshFromPolyData` module somewhere? Or, could we add a similar script to the existing repo? We've also discussed [`ItkVtkGlue`](https://github.com/InsightSoftwareConsortium/ITKVtkGlue) as a potential destination.

**Comment by dzenanz** [comment]
I suppose [this](https://github.com/InsightSoftwareConsortium/ITKApps/blob/master/Auxiliary/vtk/vtkPolyDataToITKMesh.cxx) could be revived. If it is to be added to `ITKMeshToPolyData`, then maybe it could be renamed into `ITKMeshPolyData` or something else? Not necessary.

**Comment by tbirdso** [comment]
It certainly looks like the same process. What was the reason for [ITKApps](https://github.com/InsightSoftwareConsortium/ITKApps) being archived?

**Comment by dzenanz** [comment]
Maintenance burden.

File: .github/workflows/build-test-package.yml [hunk 4]
```diff
@@ -172,6 +172,9 @@ jobs:
         python -m pip install --upgrade pip
         python -m pip install ninja
         python -m pip install itk==5.2rc2
+        python -m pip install itk-elastix
+        python -m pip install vtk
```
Comment (thewtex) [inline .github/workflows/build-test-package.yml:176]: `vtk` does not have Python 3.9 support, and we cannot add it as a dependency.

File: src/hasi/hasi/meshtomeshregistrar.py [hunk 5]
```diff
@@ -72,6 +74,58 @@ def mesh_to_image(self, mesh:MeshType, reference_image:ImageType=None) -> ImageT
...
+        # Set points directly
+        target_pts = vtk.vtkPoints()
+        target_pts.SetNumberOfPoints(target_mesh.GetNumberOfPoints())
+        for i in range(0, target_pts.GetNumberOfPoints()):
+            target_pts.SetPoint(i, target_mesh.GetPoint(i))
+        target_pd = vtk.vtkPolyData()
+        target_pd.SetPoints(target_pts)
+
+        # Do stuff in VTK
+        loc = vtk.vtkKdTreePointLocator()
```
Comment (thewtex) [inline src/hasi/hasi/meshtomeshregistrar.py:115]: Can we use [this](https://itk.org/Doxygen/html/classitk_1_1Statistics_1_1KdTree.html) here?

**Review by thewtex — CHANGES_REQUESTED** [review]

File: .github/workflows/build-test-package.yml (hunk 4 above)
Comment (tbirdso) [reply .github/workflows/build-test-package.yml:176]: Not sure I follow. This particular check `test-python-hasi-package` relies only on Python 3.7. If HASI Python scripts must support Python 3.6 - 3.9 does that mean that we cannot use the `vtk` Python package at all?

File: src/hasi/hasi/meshtomeshregistrar.py (hunk 5 above)
Comment (tbirdso) [reply src/hasi/hasi/meshtomeshregistrar.py:115]: I'm not sure I understand the benefits of [`KdTree`](https://itk.org/Doxygen/html/classitk_1_1Statistics_1_1KdTree.html) over [`vtkKdTree`](https://vtk.org/doc/nightly/html/classvtkKdTree.html). It seems like both generate a k-d tree decomposition and then search nearest neighbors. `vtkKdTree` builds from `vtkDataSet`s, does `KdTree` not share this requirement?

Comment (thewtex) [reply src/hasi/hasi/meshtomeshregistrar.py:115]: Yes, this is to keep our package dependencies down.

File: .github/workflows/build-test-package.yml (hunk 4 above)
Comment (tbirdso) [reply .github/workflows/build-test-package.yml]: Removed `vtk` package dependency.

File: src/hasi/hasi/meshtomeshregistrar.py (hunk 5 above)
Comment (tbirdso) [reply src/hasi/hasi/meshtomeshregistrar.py]: Refactored to rely on `itk::KdTree`.

**Comment by tbirdso** [comment]
Rebased on `master`.

**Pull request merged**